common.c

```diff
@@ -177,4 +177,65 @@ bool parseMovedReply(const char *str, NodeAddr *addr)
     /* Handle current or cluster-style -MOVED replies. */
     const char *p = strrchr(str, ' ');
     return NodeAddrParse(p + 1, strlen(p), addr);
+}
+
+/* future types should remain 4 characters or stack array in common.c/joinLinkIdleCallback() has to expand */
+char * join_type = "join";
+char * link_type = "link";
+
+/* Invoked when the connection is not connected or actively attempting
+ * a connection.
+ */
+void joinLinkIdleCallback(Connection *conn)
+{
+    char err_msg[42];
+
+    RedisRaftCtx *rr = ConnGetRedisRaftCtx(conn);
+    JoinLinkState *state = ConnGetPrivateData(conn);
+
+    time_t now;
+    time(&now);
+
+    if (state->failed) {
+        LOG_ERROR("Cluster %s: unrecoverable error, check logs", state->type);
+        goto exit_fail;
+    }
+
+    if (difftime(now, state->start) > rr->config->join_timeout) {
```

Comment:
We need to either have a different timeout setting or document that the join applies to shardgroup linking as well.

Reply:
I think documenting is fine, they are both "join" type operations